### ISSUE TITLE
Calling io.emit("event") with events that has no arguments result in TypeScript errors

### ISSUE THREAD

**Issue by lts20050703** [socketio/socket.io#4914]
**Describe the bug**
Calling io.emit("event") with events that either no argument result in TypeScript errors.

**To Reproduce**

Socket.IO server version: `4.7.3`

*Server*

```ts
import { Server } from "socket.io";

interface ServerToClientEvents {
  noArg: () => void;
}

const io = new Server<{}, ServerToClientEvents>();

io.emit("noArg"); // <- Error
```

**Expected behavior**
No errors

**Actual behavior**
`Argument of type 'string' is not assignable to parameter of type 'never'. ts(2345)`

**Platform:**
 - Device: Lenovo IdeaPad 5 14ITL05
 - OS: Windows 11 Home 23H2 (22631.2861)

**Additional context**
Seems to be a regression, after downgrading to 4.7.2 the error disappeared.

**Comment by lts20050703** [comment]
This seems to be caused by f6ef267b035a4db49b7d0fce438ca5c5b686f547

**Comment by ZachHaber** [comment]
@lts20050703 - from https://github.com/socketio/socket.io/issues/4813#issuecomment-1717492560

> I think most of the issues come from the fact that calling .timeout() is (at least currently) mandatory on the server side, since there is no default timeout value (the client has an ackTimeout option though).
> 
> So, for example:
> ```js
> // INVALID (will time out directly anyway)
> io.emit('withAck', 'hello', (...args) => { });
> // VALID
> io.timeout(1).emit('withAck', 'hello', (...args) => { });
> ```

For emitting with ack from the server, it is correctly an error now, because it wouldn't do what you'd expect from the beginning and would timeout immediately as [darrachequesne](https://github.com/darrachequesne) pointed out. Plus the type definition was misleading for it in 4.7.2 anyway. `e` would actually be a `number[]` rather than a `number` - if it was valid at all.

> `io.emit("withAck", "4", (e) => {}); // <- Error`

I will however look into the no-callback one, since that should be valid.

**Comment by ZachHaber** [comment]
@lts20050703, PR is up with the fix for the noArgs issue.

Until the fix is merged in and a new version is cut, you can use [patch-package](https://www.npmjs.com/package/patch-package) (or something similar) to update the `dist/typed-events.d.ts` file.


```diff
- export declare type EventNamesWithoutAck<Map extends EventsMap, K extends EventNames<Map> = EventNames<Map>> = IfAny<Last<Parameters<Map[K]>> | Map[K], K, K extends (Last<Parameters<Map[K]>> extends (...args: any[]) => any ? never : K) ? K : never>;
+ export declare type EventNamesWithoutAck<Map extends EventsMap, K extends EventNames<Map> = EventNames<Map>> = IfAny<Last<Parameters<Map[K]>> | Map[K], K, K extends (Parameters<Map[K]> extends never[] ? K : never) ? K : K extends (Last<Parameters<Map[K]>> extends (...args: any[]) => any ? never : K) ? K : never>;
```

I believe it would be line 32, but I don't have the current version installed anywhere to double check that line number (also, the line might not be exactly the same, but you are replacing the declaration of EventNamesWithoutAck).

**Comment by darrachequesne** [comment]
This should be fixed by https://github.com/socketio/socket.io/commit/cb6d2e02aa7ec03c2de1817d35cffa1128b107ef, included in version [4.7.4](https://github.com/socketio/socket.io/releases/tag/4.7.4).

Could you please check?
